### DISCUSSION TITLE
FCN8S is 2x slower when using cmake with MxNet v1.4.0

### DISCUSSION THREAD

**Discussion by leonid-pishchulin** [apache/incubator-mxnet#14817] [Q&A]
Problem: running out-of-the-box FCN8S symbol in cmake-built MxNet v1.4.0 results into 2x slow-down compared to using MxNet v1.4.0 built with config.mk 

setup: Ubuntu 16.04, CUDA 10.0, cuDNN 7.5.0, GPU Tesla T4, MxNet v1.4.0 from GitHub 

steps to re-produce: 

1) build MxNet with cmake 
git clone --recursive https://github.com/apache/incubator-mxnet.git --branch v1.4.x mxnet-v1.4.0-cmake 
cd mxnet-v1.4.0-cmake && mkdir lib && cd lib 
cmake .. -DUSE_LAPACK:bool=OFF -DUSE_MKL_IF_AVAILABLE:bool=OFF -DUSE_OPERATOR_TUNING:bool=OFF -DUSE_CPP_PACKAGE:bool=ON -DUSE_OPENCV:bool=ON -DBUILD_CPP_EXAMPLES:BOOL=OFF -DBLAS=open
make -j32 
cd - 
cd python/mxnet && ln -s ../../lib/libmxnet.so && cd - 

2) build MxNet using config.mk 
git clone --recursive https://github.com/apache/incubator-mxnet.git --branch v1.4.x mxnet-v1.4.0 
cd mxnet-v1.4.0 
copy-paste the following into config.mk 
USE_OPENCV = 1 
USE_BLAS = openblas 
USE_CUDA = 1 
USE_CUDA_PATH = /usr/local/cuda 
USE_CUDNN = 1 
make -j32 
cd python/mxnet && ln -s ../../lib/libmxnet.so && cd - 

3) download and unzip attached run_model.zip, run run_model.py
[run_model.zip](https://github.com/apache/incubator-mxnet/files/3123376/run_model.zip)

mxnet=mxnet-v1.4.0-cmake MXNET_CUDNN_AUTOTUNE_DEFAULT=1 LD_LIBRARY_PATH=/workplace/leopis/code/src/$mxnet/lib:/usr/local/cuda/lib64 PYTHONPATH=/workplace/leopis/code/src/$mxnet/python:/workplace/leopis/code/src/$mxnet/example/fcn-xs:$PWD python run_model.py 

MxNet: /workplace/leopis/code/src/mxnet-v1.4.0-cmake/python/mxnet/__init__.pyc 1.4.0 
load symbol_fcnxs 
dry run MXNet 
[16:05:37] /workplace/leopis/code/src/mxnet-v1.4.0/src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable) 
[16:05:48] /workplace/leopis/code/src/mxnet-v1.4.0/src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable) 
run MXNet 
inputs: (16 3 320 240) 
outputs: (16L, 21L, 320L, 240L) 
forward time: 55.95 ms/frame, 57288.96 ms total; batch_size: 16; symbol: symbol_fcnxs 

mxnet=mxnet-v1.4.0 MXNET_CUDNN_AUTOTUNE_DEFAULT=1 LD_LIBRARY_PATH=/workplace/leopis/code/src/$mxnet/lib:/usr/local/cuda/lib64 PYTHONPATH=/workplace/leopis/code/src/$mxnet/python:/workplace/leopis/code/src/$mxnet/example/fcn-xs:$PWD python run_model.py 

MxNet: /workplace/leopis/code/src/mxnet-v1.4.0/python/mxnet/__init__.pyc 1.4.0 
load symbol_fcnxs 
dry run MXNet 
[16:08:57] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable) 
[16:09:07] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable) 
run MXNet 
inputs: (16 3 320 240) 
outputs: (16L, 21L, 320L, 240L) 
forward time: 25.68 ms/frame, 26293.41 ms total; batch_size: 16; symbol: symbol_fcnxs

[a comment by an automated account is not shown]

**Comment by leonid-pishchulin** [comment]
I did some initial profiling with nvprof and uploaded both profiler outputs 

[mxnet-v1.4.0-profiling.zip](https://github.com/apache/incubator-mxnet/files/3123379/mxnet-v1.4.0-profiling.zip)

as can be seen from mxnet-v1.4.0-cmake.nvvp, 50% of the time is spent onto mshadow::cuda::Softmax3DKernel and mshadow::cuda::MapPlanLargeKernel 
excluding both kernel runs from the overall run-time, we would get into the range of mxnet-v1.4.0 built with make

**Comment by KellenSunderland** [comment]
Thanks for the report @leonid-pishchulin.  I ran the script you provided on my home machine (RTX 2070) and got some interesting results:

```
MXNet with make
# forward time: 15.66 ms/frame, 16043.96 ms total; batch_size: 16; symbol: symbol_fcnxs

MXNet with cmake:
# forward time: 15.88 ms/frame, 16261.01 ms total; batch_size: 16; symbol: symbol_fcnxs
```

So at least on my environment it seems that both run approximately the same speed.  However that's not to say that on your environment (and possibly in the case of many of our users) something is regressing.

**Comment by KellenSunderland** [comment]
As a next step I'd recommend we try compiling with a Dockerfile to make sure we are using the same compiler versions and environment.  Would it also be possible to post the output of running cuobjdump in gist.github.com and linking it here for both .sos?

**Comment by leonid-pishchulin** [comment]
commenting out softmax layer reduces run-time from 55.95 to 27.25 ms / fr in case of MXNet with cmake

MXNet with make
forward time: 27.66 ms/frame, 28320.44 ms total; batch_size: 16; symbol: symbol_fcnxs

MXNet with cmake:
forward time: 27.25 ms/frame, 27902.43 ms total; batch_size: 16; symbol: symbol_fcnxs
